core/network-libp2p/src/custom_proto/handler.rs

```diff
@@ -165,44 +158,199 @@ struct PerProtocolNormalState<TMessage, TSubstream> {
+	/// Polls for things that are new. Same API constraints as `Future::poll()`.
+	/// Optionally returns the event to produce.
+	/// You must pass the `protocol_id` as we need have to inject it in the returned event.
+	/// API note: Ideally we wouldn't need to be passed a `ProtocolId`, and we would return a
+	/// 	different enum that doesn't contain any `protocol_id`, and the caller would inject
+	/// 	the ID itself, but that's a ton of code for not much gain.
+	fn poll(&mut self) -> Option<CustomProtoHandlerOut<TMessage>> {
+		for n in (0..self.pending_response.len()).rev() {
+			let (request_id, mut substream) = self.pending_response.swap_remove(n);
+			match substream.poll() {
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Message(message)))) => {
+					if message.request_id() == CustomMessageId::Response(request_id) {
+						let event = CustomProtoHandlerOut::CustomMessage {
+							message
+						};
+						self.shutdown.push(substream);
+						return Some(event);
+					} else {
+						self.shutdown.push(substream);
+						let event = CustomProtoHandlerOut::ProtocolError {
+							is_severe: true,
+							error: format!("Request ID doesn't match substream: expected {:?}, \
+								got {:?}", request_id, message.request_id()).into(),
+						};
+						return Some(event);
+					}
+				},
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Clogged { .. }))) =>
+					unreachable!("Cannot receive Clogged message with new protocol version; QED"),
+				Ok(Async::NotReady) =>
+					self.pending_response.push((request_id, substream)),
+				Ok(Async::Ready(None)) => {
+					self.shutdown.push(substream);
+					let event = CustomProtoHandlerOut::ProtocolError {
+						is_severe: false,
+						error: format!("Request ID {:?} didn't receive an answer", request_id).into(),
+					};
+					return Some(event);
+				}
+				Err(err) => {
+					self.shutdown.push(substream);
+					let event = CustomProtoHandlerOut::ProtocolError {
+						is_severe: false,
+						error: format!("Error while waiting for an answer for {:?}: {}",
+							request_id, err).into(),
+					};
+					return Some(event);
+				}
+			}
+		}
 
-		let return_value;
+		for n in (0..self.incoming_substreams.len()).rev() {
+			let mut substream = self.incoming_substreams.swap_remove(n);
+			match substream.poll() {
+				Ok(Async::Ready(Some(RegisteredProtocolEvent::Message(message)))) => {
+					if let CustomMessageId::Request(id) = message.request_id() {
+						self.pending_send_back.push((id, substream));
+						return Some(CustomProtoHandlerOut::CustomMessage {
+							message
+						});
+					} else if let CustomMessageId::OneWay = message.request_id() {
+						self.shutdown.push(substream);
+						return Some(CustomProtoHandlerOut::CustomMessage {
+							message
+						});
+					} else {
+						self.shutdown.push(substream);
+						return Some(CustomProtoHandlerOut::ProtocolError {
```

Comment:
You could return the entire `if-else` expression, would be clearer that this branch always returns.